### PR TITLE
[Balance] Checklist

### PR DESCRIPTION
**Pull request by Gebuz**
- Added Checklist
- Added Cancelled Casts
- Merged Moon spells in statistic boxes and suggestions
- Temporary fix to Moon spells in cast efficiency (needs to be merged into one spell to work properly because of charges)
- refactor of some classes to support the Checklist

### CODE REVIEW

File: src/Parser/Druid/Balance/Modules/Features/CancelledCasts.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  suggestions(when) {
+    const cancelledPercentage = this.castsCancelled / this.totalCasts;
+
+    when(cancelledPercentage).isGreaterThan(0.05)
+      .addSuggestion((suggest, actual, recommended) => {
+        return suggest(<span>You cancelled {formatPercentage(cancelledPercentage)}% of your spells. While it is expected that you will have to cancel a few casts to react to a boss mechanic or to move, you should try to ensure that you are cancelling as few casts as possible.</span>)
+          .icon('inv_misc_map_01')
+          .actual(`${formatPercentage(actual)}% casts cancelled`)
+          .recommended(`<${formatPercentage(recommended)}% is recommended`)
+          .regular(.1).major(recommended + 0.2);
```
Comment (kfinch) [inline src/Parser/Druid/Balance/Modules/Features/CancelledCasts.js]: `regular` being absolute and `major` being relative is a little awkward...

File: src/Parser/Druid/Balance/Modules/Features/Checklist.js [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+      },
+    }),
+    new Rule({
+      name: 'Maintain your DoTs on the boss',
+      description: 'DoTs are a big part of your damage. You should try to keep as high uptime on them as possible.',
+      requirements: () => {
+        const combatant = this.combatants.selected;
+        return [
+          new Requirement({
+            name: <Wrapper><SpellLink id={SPELLS.MOONFIRE_BEAR.id} icon/> uptime</Wrapper>,
```
Comment (kfinch) [inline src/Parser/Druid/Balance/Modules/Features/Checklist.js]: Moonfire... Bear? Have you tested this and it's the right spellId? I'm guessing this is like "Guardian got to the spellId first" or something...

File: src/Parser/Druid/Balance/Modules/Features/StellarFlareUptime.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    const stellarFlareUptime = this.enemies.getBuffUptime(SPELLS.STELLAR_FLARE_TALENT.id) / this.owner.fightDuration;
+    return {
+      actual: stellarFlareUptime,
+      isLessThan: {
+        minor: 0.95,
+        average: 0.9,
+        major: 0.8,
+      },
+      style: 'percentage',
+      text: 'Your Stellar Flare uptime can be improved. Try to pay more attention to your Moonfire on the boss.',
```
Comment (kfinch) [inline src/Parser/Druid/Balance/Modules/Features/StellarFlareUptime.js]: `pay more attention to your Moonfire` should be `pay more attention to your Stellar Flare`

File: src/Parser/Druid/Balance/Modules/Features/SunfireUptime.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    const sunfireUptime = this.enemies.getBuffUptime(SPELLS.SUNFIRE.id) / this.owner.fightDuration;
+    return {
+      actual: sunfireUptime,
+      isLessThan: {
+        minor: 0.95,
+        average: 0.9,
+        major: 0.8,
+      },
+      style: 'percentage',
+      text: 'Your Sunfire uptime can be improved. Try to pay more attention to your Moonfire on the boss.',
```
Comment (kfinch) [inline src/Parser/Druid/Balance/Modules/Features/SunfireUptime.js]: `pay more attention to your Moonfire` should be `pay more attention to your Sunfire`

File: src/Parser/Druid/Balance/Modules/Features/SunfireUptime.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+      });
+  }
+
+  statistic() {
+    const sunfireUptime = this.enemies.getBuffUptime(SPELLS.SUNFIRE.id) / this.owner.fightDuration;
+    return (
+      <StatisticBox
+        icon={<SpellIcon id={SPELLS.SUNFIRE.id} />}
+        value={`${formatPercentage(sunfireUptime)} %`}
+        label="sunfire uptime"
```
Comment (kfinch) [inline src/Parser/Druid/Balance/Modules/Features/SunfireUptime.js]: First letter of statistic label should be capitalized.

File: src/Parser/Druid/Balance/Modules/Features/Checklist.js (hunk 2 above)
Comment (Gebuz) [reply src/Parser/Druid/Balance/Modules/Features/Checklist.js]: Yep, I tried the other one first, but only the bear one works. I don't know what the other one is for. this is the only comments I could find about it:
  // MOONFIRE = MOONFIRE_BEAR
  // MOONFIRE_CAST = MOONFIRE

Comment (kfinch) [reply src/Parser/Druid/Balance/Modules/Features/Checklist.js]: Oh, the Moonfire debuff id is called `MOONFIRE_BEAR` and the Moonfire cast id is called `MOONFIRE`. That should really be refactored, but it's not your job to do it here.

**Pull request merged**